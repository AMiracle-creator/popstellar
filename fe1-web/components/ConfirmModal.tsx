import React from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet, View, ViewStyle, Modal, TextStyle, Text
} from 'react-native';
import STRINGS from 'res/strings';
import { white } from 'styles/colors';
import { Typography } from 'styles';
import WideButtonView from './WideButtonView';

const styles = StyleSheet.create({
  modalView: {
    backgroundColor: white,
    borderRadius: 10,
    borderWidth: 1,
    margin: 'auto',
    width: 550,
  } as ViewStyle,
  titleView: {
    borderBottomWidth: 1,
  } as ViewStyle,
  modalTitle: {
    ...Typography.important,
    alignSelf: 'flex-start',
    padding: 20,
    paddingLeft: 10,
  } as TextStyle,
  modalDescription: {
    ...Typography.base,
    fontSize: 20,
    alignSelf: 'flex-start',
    textAlign: 'left',
    padding: 20,
    paddingLeft: 10,
  } as TextStyle,
  buttonView: {
    alignSelf: 'center',
    flexDirection: 'row',
    paddingBottom: 20,
  } as ViewStyle,
});

const ConfirmModal = (props: IPropTypes) => {
  const { visibility } = props;
  const { setVisibility } = props;
  const { title } = props;
  const { description } = props;
  const { buttonConfirmText } = props;
  const { buttonCancelText } = props;
  const { onConfirmPress } = props;

  return (
    <Modal
      visible={visibility}
      onRequestClose={() => setVisibility(!visibility)}
      transparent
    >
      <View style={styles.modalView}>
        <View style={styles.titleView}>
          <Text style={styles.modalTitle}>{title}</Text>
        </View>
        <Text style={styles.modalDescription}>{description}</Text>
        <View style={styles.buttonView}>
          <WideButtonView
            title={buttonConfirmText}
            onPress={() => onConfirmPress()}
          />
          <WideButtonView
            title={buttonCancelText}
            onPress={() => setVisibility(!visibility)}
          />
        </View>
      </View>
    </Modal>
  );
};

const propTypes = {
  visibility: PropTypes.bool.isRequired,
  setVisibility: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  buttonCancelText: PropTypes.string,
  buttonConfirmText: PropTypes.string,
  onConfirmPress: PropTypes.func.isRequired,
};

ConfirmModal.propTypes = propTypes;

ConfirmModal.defaultProps = {
  buttonCancelText: STRINGS.general_button_cancel,
  buttonConfirmText: STRINGS.general_button_confirm,
};

type IPropTypes = {
  visibility: boolean,
  setVisibility: Function,
  title: string,
  description: string,
  buttonCancelText: string,
  buttonConfirmText: string,
  onConfirmPress: Function,
};

export default ConfirmModal;
