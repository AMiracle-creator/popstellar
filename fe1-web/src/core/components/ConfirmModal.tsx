import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { Modal, Text } from 'react-native';
import { ScrollView, TouchableWithoutFeedback } from 'react-native-gesture-handler';

import { ModalStyles, Typography } from 'core/styles';
import STRINGS from 'resources/strings';

import Button from './Button';
import Input from './Input';
import ModalHeader from './ModalHeader';

/**
 * A modal used to ask for the confirmation or cancellation of the user.
 * It can also ask for a user input.
 */

const ConfirmModal = (props: IPropTypes) => {
  const {
    visibility,
    setVisibility,
    title,
    description,
    buttonConfirmText,
    onConfirmPress,
    hasTextInput,
    textInputPlaceholder,
  } = props;

  const [textInput, setTextInput] = useState('');

  return (
    <Modal
      transparent
      visible={visibility}
      onRequestClose={() => {
        setVisibility(!visibility);
      }}>
      <TouchableWithoutFeedback
        containerStyle={ModalStyles.modalBackground}
        onPress={() => {
          setVisibility(!visibility);
        }}
      />
      <ScrollView style={ModalStyles.modalContainer}>
        <ModalHeader onClose={() => setVisibility(!visibility)}>{title}</ModalHeader>
        <Text style={Typography.paragraph}>{description}</Text>
        {hasTextInput ? (
          <Input
            testID="confirm-modal-input"
            value={textInput}
            onChange={setTextInput}
            placeholder={textInputPlaceholder}
            border
          />
        ) : null}
        <Button onPress={() => onConfirmPress(textInput)} testID="confirm-modal-confirm">
          <Text style={[Typography.base, Typography.centered, Typography.negative]}>
            {buttonConfirmText}
          </Text>
        </Button>
      </ScrollView>
    </Modal>
  );
};

const propTypes = {
  visibility: PropTypes.bool.isRequired,
  setVisibility: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  buttonConfirmText: PropTypes.string,
  onConfirmPress: PropTypes.func.isRequired,
  hasTextInput: PropTypes.bool,
  textInputPlaceholder: PropTypes.string,
};

ConfirmModal.propTypes = propTypes;

ConfirmModal.defaultProps = {
  buttonConfirmText: STRINGS.general_button_confirm,
  hasTextInput: false,
  textInputPlaceholder: '',
};

type IPropTypes = {
  visibility: boolean;
  setVisibility: Function;
  title: string;
  description: string;
  buttonConfirmText: string;
  onConfirmPress: Function;
  hasTextInput: boolean;
  textInputPlaceholder: string;
};

export default ConfirmModal;
