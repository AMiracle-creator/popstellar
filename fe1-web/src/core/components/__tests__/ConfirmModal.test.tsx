import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import ConfirmModal from '../ConfirmModal';

const TITLE = 'Title';
const DESCRIPTION = 'Description';
const CONFIRM = 'Confirm';

let setModalIsVisible: Function;
let onConfirmPress: Function;

beforeEach(() => {
  setModalIsVisible = jest.fn();
  onConfirmPress = jest.fn();
});

describe('ConfirmModal', () => {
  it('renders correctly', () => {
    const component = render(
      <ConfirmModal
        visibility
        setVisibility={setModalIsVisible}
        title={TITLE}
        description={DESCRIPTION}
        onConfirmPress={() => onConfirmPress()}
      />,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with text input', () => {
    const component = render(
      <ConfirmModal
        visibility
        setVisibility={setModalIsVisible}
        title={TITLE}
        description={DESCRIPTION}
        onConfirmPress={() => onConfirmPress()}
        hasTextInput
      />,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('disappears correctly after dismissing', () => {
    const { getByTestId, toJSON } = render(
      <ConfirmModal
        visibility
        setVisibility={setModalIsVisible}
        title={TITLE}
        description={DESCRIPTION}
        onConfirmPress={() => onConfirmPress()}
      />,
    );
    const cancelButton = getByTestId('modal-header-close');
    fireEvent.press(cancelButton);
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onConfirmPress correctly', () => {
    const { getByTestId } = render(
      <ConfirmModal
        visibility
        setVisibility={setModalIsVisible}
        title={TITLE}
        description={DESCRIPTION}
        onConfirmPress={() => onConfirmPress()}
        buttonConfirmText={CONFIRM}
      />,
    );
    const confirmButton = getByTestId('confirm-modal-confirm');
    fireEvent.press(confirmButton);
    expect(onConfirmPress).toHaveBeenCalledTimes(1);
  });

  it('calls onConfirmPress with textInput', () => {
    const mockInput = 'input';

    const { getByTestId } = render(
      <ConfirmModal
        visibility
        setVisibility={setModalIsVisible}
        title={TITLE}
        description={DESCRIPTION}
        onConfirmPress={onConfirmPress}
        buttonConfirmText={CONFIRM}
        hasTextInput
      />,
    );
    const textInput = getByTestId('confirm-modal-input');
    fireEvent.changeText(textInput, mockInput);

    const confirmButton = getByTestId('confirm-modal-confirm');
    fireEvent.press(confirmButton);
    expect(onConfirmPress).toHaveBeenCalledWith(mockInput);
  });
});
