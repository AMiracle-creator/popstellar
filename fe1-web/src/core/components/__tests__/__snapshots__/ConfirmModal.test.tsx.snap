// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal disappears correctly after dismissing 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 10,
        "borderWidth": 1,
        "margin": "auto",
        "width": 550,
      }
    }
  >
    <View
      style={
        Object {
          "borderBottomWidth": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "flex-start",
            "fontSize": 20,
            "fontWeight": "bold",
            "lineHeight": 26,
            "marginHorizontal": 10,
            "padding": 20,
            "paddingLeft": 10,
            "textAlign": "center",
          }
        }
      >
        Title
      </Text>
    </View>
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontSize": 20,
          "lineHeight": 26,
          "marginHorizontal": 10,
          "padding": 20,
          "paddingLeft": 10,
          "textAlign": "left",
        }
      }
    >
      Description
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
          "flexDirection": "row",
          "paddingBottom": 20,
        }
      }
    >
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        style={
          Object {
            "marginBottom": 16,
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#3742fa",
                "borderColor": "#3742fa",
                "borderRadius": 16,
                "borderWidth": 1,
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "lineHeight": 26,
                    "textAlign": "left",
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#fff",
                  },
                ]
              }
            >
              Confirm
            </Text>
          </View>
        </View>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        style={
          Object {
            "marginBottom": 16,
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#3742fa",
                "borderColor": "#3742fa",
                "borderRadius": 16,
                "borderWidth": 1,
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "lineHeight": 26,
                    "textAlign": "left",
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#fff",
                  },
                ]
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </RNGestureHandlerButton>
    </View>
  </View>
</Modal>
`;

exports[`ConfirmModal renders correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  presentationStyle="formSheet"
  transparent={true}
  visible={true}
>
  <RNGestureHandlerButton
    collapsable={false}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.25)",
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "padding": 32,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      nativeID="animatedComponent"
      style={Object {}}
    />
  </RNGestureHandlerButton>
  <RCTScrollView
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 16,
        "margin": 32,
        "padding": 32,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "color": "#000",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 26,
              "textAlign": "left",
            }
          }
        >
          Title
        </Text>
        <RNGestureHandlerButton
          collapsable={false}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          style={
            Object {
              "position": "absolute",
              "right": 0,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            nativeID="animatedComponent"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#3742fa",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "lineHeight": 26,
                  "textAlign": "left",
                }
              }
            >
              Done
            </Text>
          </View>
        </RNGestureHandlerButton>
      </View>
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 20,
            "lineHeight": 26,
            "marginBottom": 16,
            "textAlign": "left",
          }
        }
      >
        Description
      </Text>
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        style={
          Object {
            "marginBottom": 16,
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#3742fa",
                  "borderColor": "#3742fa",
                  "borderRadius": 16,
                  "borderWidth": 1,
                  "padding": 8,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 20,
                    "lineHeight": 26,
                    "textAlign": "left",
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#fff",
                  },
                ]
              }
            >
              Confirm
            </Text>
          </View>
        </View>
      </RNGestureHandlerButton>
    </View>
  </RCTScrollView>
</Modal>
`;

exports[`ConfirmModal renders correctly with text input 1`] = `
<Modal
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  presentationStyle="formSheet"
  transparent={true}
  visible={true}
>
  <RNGestureHandlerButton
    collapsable={false}
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.25)",
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "padding": 32,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      nativeID="animatedComponent"
      style={Object {}}
    />
  </RNGestureHandlerButton>
  <RCTScrollView
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 16,
        "margin": 32,
        "padding": 32,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "color": "#000",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 26,
              "textAlign": "left",
            }
          }
        >
          Title
        </Text>
        <RNGestureHandlerButton
          collapsable={false}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          style={
            Object {
              "position": "absolute",
              "right": 0,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            nativeID="animatedComponent"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#3742fa",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "lineHeight": 26,
                  "textAlign": "left",
                }
              }
            >
              Done
            </Text>
          </View>
        </RNGestureHandlerButton>
      </View>
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 20,
            "lineHeight": 26,
            "marginBottom": 16,
            "textAlign": "left",
          }
        }
      >
        Description
      </Text>
      <TextInput
        editable={true}
        onChangeText={[Function]}
        placeholder=""
        style={
          Array [
            Object {
              "color": "#000",
              "fontSize": 20,
              "lineHeight": 26,
              "marginBottom": 16,
              "textAlign": "left",
            },
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 8,
              "flex": 1,
              "padding": 8,
            },
            Object {
              "borderColor": "#000",
              "borderWidth": 1,
            },
          ]
        }
        value=""
      />
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        style={
          Object {
            "marginBottom": 16,
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#3742fa",
                  "borderColor": "#3742fa",
                  "borderRadius": 16,
                  "borderWidth": 1,
                  "padding": 8,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 20,
                    "lineHeight": 26,
                    "textAlign": "left",
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#fff",
                  },
                ]
              }
            >
              Confirm
            </Text>
          </View>
        </View>
      </RNGestureHandlerButton>
    </View>
  </RCTScrollView>
</Modal>
`;
